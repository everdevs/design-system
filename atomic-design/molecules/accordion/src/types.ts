import { PropsWithTheme } from "@evernest/theme";

export type AccordionElement = HTMLDivElement;

export interface AccordionProps extends PropsWithTheme {
	title: string;
	headerComponent?: keyof JSX.IntrinsicElements | React.ComponentType<any>;
}

export interface StyledAccordionProps extends PropsWithTheme<AccordionElement> {}

export interface StyledButtonProps extends PropsWithTheme<HTMLButtonElement> {}

export interface StyledPanelProps extends PropsWithTheme<AccordionElement> {
	expanded: boolean;
}
