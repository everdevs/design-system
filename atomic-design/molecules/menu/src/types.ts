import { PropsWithTheme } from "@evernest/theme";

export type MenuElement = HTMLDivElement;

export enum MenuDirection {
	vertical = "vertical",
	horizontal = "horizontal",
}

export interface StyledMenuProps extends PropsWithTheme<MenuElement> {
	/** Add types */
	direction?: MenuDirection;
}
