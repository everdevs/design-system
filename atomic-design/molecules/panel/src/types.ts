import { PropsWithTheme } from "@evernest/theme";

export type PanelElement = HTMLDivElement;
export type ToggleElement = HTMLButtonElement;

export enum Anchor {
	bottom = "bottom",
	left = "left",
	right = "right",
	top = "top",
}

export interface StyledPanelProps extends PropsWithTheme<PanelElement> {
	/** Add types */
	anchor: Anchor;
	onClose?(): void;
	open?: boolean;
	width?: string;
}

export interface StyledBackdropProps extends PropsWithTheme {
	/** Add types */
	open?: boolean;
}

export interface PanelProps extends PanelElement {
	/** Add types */
}

export interface StyledToggleProps extends PropsWithTheme<ToggleElement> {
	/** Add types */
	open?: boolean;
}

export interface ToggleProps extends ToggleElement {
	/** Add types */
}
