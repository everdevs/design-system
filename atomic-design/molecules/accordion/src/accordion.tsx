import { css } from "@emotion/core";
import styled from "@emotion/styled";
import { Icon, Size } from "@evernest/icon";
import { PropsWithTheme } from "@evernest/theme";
import React from "react";
import { animated, useSpring } from "react-spring";
import { useMeasure } from "react-use";
import { AccordionElement, AccordionProps, StyledAccordionProps } from "./types";

/* @todo: revisit border style with token refactor */
export const StyledAccordion = styled("div", {
	shouldForwardProp: (propName: string) => !["theme"].includes(propName),
})<StyledAccordionProps>`
	${({ theme: { palette } }) => css`
		border-bottom: 1px solid ${palette.brightGrey.css};
	`};
`;

export const StyledButton = styled.button`
	appearance: none;
	background: none;
	border-radius: 0;
	border: none;
	color: currentColor;
	font-family: inherit;
	font-size: inherit;
	padding: 0 0 var(--spacing-xs);
	width: 100%;
`;

export const StyledInnerButtonWrapper = styled.div`
	display: flex;
	height: 100%;
	overflow: hidden;
	text-align: left;
`;

export const StyledPanel = styled.div<PropsWithTheme>`
	padding-bottom: var(--spacing-xs);
	padding-left: var(--spacing-s);
	${({ theme: { mq } }) => css`
		@media ${mq.l} {
			padding-left: var(--spacing-m);
		}
	`};
`;

export const StyledAnimatedPanelWrapper = styled(animated.div)`
	overflow: hidden;
`;

export const StyledIconWrapper = styled(animated.div)<PropsWithTheme>`
	display: inline-flex;
	min-width: var(--spacing-s);
	${({ theme: { mq } }) => css`
		@media ${mq.l} {
			min-width: var(--spacing-m);
		}
	`};
`;

export const Accordion = React.forwardRef<AccordionElement, AccordionProps>(
	({ id, children, title, headerComponent, springConfig, ...props }, ref) => {
		const [expanded, setExpanded] = React.useState(false);
		const [panelBottomPadding, setPanelBottomPadding] = React.useState(0);

		const handleClick = () => setExpanded(!expanded);

		const panelId = `${id}-panel`;

		const [useMeasureRef, { height }] = useMeasure<HTMLDivElement>();

		const springPanelProps = useSpring({
			config: springConfig,
			height: expanded ? height + panelBottomPadding : 0,
		});

		const { z } = useSpring({
			config: springConfig,
			z: expanded ? 0.125 : 0,
		});

		React.useEffect(() => {
			const panelEl = document.getElementById(panelId);
			/* istanbul ignore next */
			if (panelEl) {
				const paddingBottom = getComputedStyle(panelEl).paddingBottom;
				setPanelBottomPadding(parseInt(paddingBottom));
			}
		}, [panelId, setPanelBottomPadding]);

		const HeaderComponent = React.useMemo(() => headerComponent, [headerComponent]);

		return (
			<StyledAccordion {...props} ref={ref}>
				<HeaderComponent>
					<StyledButton
						aria-controls={panelId}
						aria-expanded={expanded}
						id={id}
						onClick={handleClick}
					>
						<StyledInnerButtonWrapper>
							<StyledIconWrapper>
								<animated.span
									style={{
										transform: z.interpolate(
											value => `rotate3d(0,0,1,${value}turn)`
										),
									}}
								>
									<Icon aria-hidden="true" icon="close" size={Size.medium} />
								</animated.span>
							</StyledIconWrapper>
							<span data-test-id="styled-inner-button-wrapper-label">{title}</span>
						</StyledInnerButtonWrapper>
					</StyledButton>
				</HeaderComponent>
				<StyledAnimatedPanelWrapper style={springPanelProps}>
					<StyledPanel
						ref={useMeasureRef}
						aria-labelledby={id}
						id={panelId}
						role="region"
					>
						{children}
					</StyledPanel>
				</StyledAnimatedPanelWrapper>
			</StyledAccordion>
		);
	}
);

Accordion.defaultProps = {
	headerComponent: "div",
	springConfig: undefined,
};
