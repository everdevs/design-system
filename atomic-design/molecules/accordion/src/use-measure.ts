import React from "react";

export interface Measurements {
	contentRect: Pick<
		DOMRectReadOnly,
		"x" | "y" | "top" | "left" | "right" | "bottom" | "height" | "width"
	>;
	borderBoxSize: ResizeObserverSize;
	contentBoxSize: ResizeObserverSize;
}

const intialState = {
	contentRect: {
		x: 0,
		y: 0,
		width: 0,
		height: 0,
		top: 0,
		right: 0,
		bottom: 0,
		left: 0,
	},
	borderBoxSize: {
		inlineSize: 0,
		blockSize: 0,
	},
	contentBoxSize: {
		inlineSize: 0,
		blockSize: 0,
	},
};

/* @todo: revisit testing when ResizeObserver is better implemented */
/* istanbul ignore next */
const useMeasure = <T extends Element = HTMLDivElement>(): [
	React.MutableRefObject<T>,
	Measurements
] => {
	const [bounds, setBounds] = React.useState<Measurements>(intialState);
	const ref = React.useRef<T>();

	React.useLayoutEffect(() => {
		let animationFrameId: number;

		const measure: ResizeObserverCallback = ([entry]) => {
			animationFrameId = window.requestAnimationFrame(() => {
				setBounds({
					contentRect: entry.contentRect,
					borderBoxSize: entry.borderBoxSize[0],
					contentBoxSize: entry.contentBoxSize[0],
				});
			});
		};

		const resizeObserver = new ResizeObserver(measure);
		resizeObserver.observe(ref.current);

		return () => {
			window.cancelAnimationFrame(animationFrameId);
			resizeObserver.disconnect();
		};
	}, []);

	return [ref, bounds];
};

export default useMeasure;
