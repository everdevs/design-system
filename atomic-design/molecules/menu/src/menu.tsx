import { css } from "@emotion/react";
import styled from "@emotion/styled";
import React from "react";
import { MenuDirection, MenuElement, StyledMenuProps } from "./types";

export const StyledMenuTpl = styled.nav<StyledMenuProps>`
	display: flex;
	${({ direction }) => css`
		flex-direction: ${direction === MenuDirection.vertical ? "column" : "row"};
	`};
`;

export const Menu = React.forwardRef<MenuElement, StyledMenuProps>((props, ref) => (
	<StyledMenuTpl {...props} ref={ref} />
));
