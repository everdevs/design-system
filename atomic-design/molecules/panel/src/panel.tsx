import styled from "@emotion/styled";
import { css, Global } from "@emotion/core";
import React from "react";
import { Anchor, PanelElement, StyledBackdropProps, StyledPanelProps } from "./types";

export const StyledBackdrop = styled.div<StyledBackdropProps>`
	position: fixed;
	z-index: 100;
	top: 0;
	right: 0;
	bottom: 0;
	left: 0;
	opacity: 0;
	visibility: hidden;
	pointer-events: none;
	transition-duration: 0.3s;
	transition-property: opacity, visibility;
	${({ theme: { palette } }) =>
		css`
			background: ${palette.black.css};
		`};
	${({ open }) =>
		open &&
		css`
			visibility: visible;
			pointer-events: all;
			opacity: 0.7;
		`};
`;

export const StyledPanel = styled.aside<StyledPanelProps>`
	position: fixed;
	z-index: 100;
	transition-duration: 0.3s;
	transition-property: transform;
	overflow: auto;
	${({ theme: { palette } }) => css`
		background: ${palette.white.css};
	`};
	${({ anchor, width, open }) => {
		switch (anchor) {
			case Anchor.bottom: {
				return css`
					bottom: 0;
					left: 0;
					right: 0;
					transform: translate3d(0, ${open ? 0 : "100%"}, 0);
				`;
			}
			case Anchor.left: {
				return css`
					bottom: 0;
					left: 0;
					top: 0;
					width: ${width};
					max-width: calc(100vw - 80px);
					transform: translate3d(${open ? 0 : "-100%"}, 0, 0);
				`;
			}
			case Anchor.right: {
				return css`
					bottom: 0;
					top: 0;
					right: 0;
					width: ${width};
					max-width: calc(100vw - 80px);
					transform: translate3d(${open ? 0 : "100%"}, 0, 0);
				`;
			}
			case Anchor.top: {
				return css`
					top: 0;
					left: 0;
					right: 0;
					transform: translate3d(0, ${open ? 0 : "-100%"}, 0);
				`;
			}
			default: {
				break;
			}
		}
	}};
`;

const blockScroll = css`
	body {
		overflow: hidden;
		height: 100vh;
	}
`;

export const Panel = React.forwardRef<PanelElement, StyledPanelProps>(
	({ open, onClose, ...props }, ref) => {
		const handleClick = React.useCallback(() => {
			open && onClose();
		}, [open, onClose]);
		return (
			<>
				{open && <Global styles={blockScroll} />}
				<StyledBackdrop open={open} onClick={handleClick} />
				<StyledPanel {...props} open={open} ref={ref} />
			</>
		);
	}
);

Panel.defaultProps = {
	width: "295px",
};
