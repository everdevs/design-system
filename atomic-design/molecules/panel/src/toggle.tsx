import styled from "@emotion/styled";
import React from "react";
import { StyledToggleProps, ToggleElement } from "./types";
import { css } from "@emotion/react";

export const StyledBurgerWrapper = styled.div`
	height: 48px;
	width: 48px;
	position: relative;
`;

export const StyledBurgerBun = styled.span`
	height: 2px;
	width: 16px;
	position: absolute;
	top: 50%;
	left: 50%;
	margin: -1px -8px;
	background: currentColor;
	transition-duration: 0.3s;
	transition-property: transform;
`;

const bunTransformation = {
	top: {
		open: "translate3d(0, 0, 0) rotate3d(0,0,1,45deg)",
		closed: "translate3d(0, -4px, 0) rotate3d(0,0,0,0deg)",
	},
	bottom: {
		open: "translate3d(0, 0, 0) rotate3d(0,0,1,-45deg)",
		closed: "translate3d(0, 4px, 0) rotate3d(0,0,0,0deg)",
	},
};
export const Burger: React.FC<StyledToggleProps> = ({ open }) => {
	const [topBun, setTopBun] = React.useState(bunTransformation.top[open ? "open" : "closed"]);
	const [bottomBun, setBottomBun] = React.useState(
		bunTransformation.bottom[open ? "open" : "closed"]
	);
	React.useEffect(() => {
		setTopBun(bunTransformation.top[open ? "open" : "closed"]);
		setBottomBun(bunTransformation.bottom[open ? "open" : "closed"]);
	}, [open]);
	return (
		<StyledBurgerWrapper>
			<StyledBurgerBun style={{ transform: topBun }} />
			<StyledBurgerBun style={{ transform: bottomBun }} />
		</StyledBurgerWrapper>
	);
};

export const StyledToggle = styled.button<StyledToggleProps>`
	position: fixed;
	top: 12px;
	right: 16px;
	z-index: 101;
	height: 48px;
	width: 48px;
	background: none;
	font-size: 1em;
	border: 0;
	border-radius: 50%;
	padding: 0;
	margin: 0;
	box-shadow: none;
	display: flex;
	${({ theme: { palette } }) => css`
		color: ${palette.white.css};
	`};
`;

export const HiddenText = styled.span`
	position: absolute;
	top: -100vh;
	left: -100vw;
	display: block;
	height: 0;
	width: 0;
	opacity: 0;
`;

export const Toggle = React.forwardRef<ToggleElement, StyledToggleProps>(
	({ children, open, ...props }, ref) => (
		<StyledToggle {...props} open={open} ref={ref}>
			<Burger open={open} />
			<HiddenText>{open ? "close" : "open"} Menu</HiddenText>
		</StyledToggle>
	)
);
