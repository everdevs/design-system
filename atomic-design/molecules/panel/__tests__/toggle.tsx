import { theme } from "@evernest/theme";
import { mount } from "enzyme";
import React from "react";
import renderer from "react-test-renderer";
import { Toggle } from "../src";

test("Toggle allows setting classname", () => {
	const className = "test";
	const wrapper = mount(<Toggle className={className} theme={theme} />);
	expect(wrapper.find(Toggle)).toHaveClassName(className);
});

test("Panel can be open or closed", () => {
	const opened = renderer.create(<Toggle open theme={theme} />).toJSON();
	const closed = renderer.create(<Toggle theme={theme} />).toJSON();
	expect(opened).toMatchSnapshot();
	expect(closed).toMatchSnapshot();
});
