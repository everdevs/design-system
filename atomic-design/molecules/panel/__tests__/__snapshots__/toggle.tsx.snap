// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel can be open or closed 1`] = `
<button
  className="css-aulye4"
  open={true}
>
  <div
    className="css-1twxhdo"
  >
    <span
      className="css-oo1k7y"
      style={
        Object {
          "transform": "translate3d(0, 0, 0) rotate3d(0,0,1,45deg)",
        }
      }
    />
    <span
      className="css-oo1k7y"
      style={
        Object {
          "transform": "translate3d(0, 0, 0) rotate3d(0,0,1,-45deg)",
        }
      }
    />
  </div>
  <span
    className="css-obzl7d"
  >
    close
     Menu
  </span>
</button>
`;

exports[`Panel can be open or closed 2`] = `
<button
  className="css-aulye4"
>
  <div
    className="css-1twxhdo"
  >
    <span
      className="css-oo1k7y"
      style={
        Object {
          "transform": "translate3d(0, -4px, 0) rotate3d(0,0,0,0deg)",
        }
      }
    />
    <span
      className="css-oo1k7y"
      style={
        Object {
          "transform": "translate3d(0, 4px, 0) rotate3d(0,0,0,0deg)",
        }
      }
    />
  </div>
  <span
    className="css-obzl7d"
  >
    open
     Menu
  </span>
</button>
`;
